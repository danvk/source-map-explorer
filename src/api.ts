import glob from 'glob';
import { partition, flatMap, isString } from 'lodash';

import { exploreBundle, UNMAPPED_KEY, SOURCE_MAP_COMMENT_KEY } from './explore';
import { AppError, getErrorMessage } from './app-error';
import {
  BundlesAndFileTokens,
  ExploreOptions,
  ExploreResult,
  Bundle,
  ExploreErrorResult,
  ExploreBundleResult,
} from './index';
import { formatOutput, saveOutputToFile } from './output';

/**
 * Analyze bundle(s)
 */
export async function explore(
  bundlesAndFileTokens: BundlesAndFileTokens,
  options: ExploreOptions = {}
): Promise<ExploreResult> {
  bundlesAndFileTokens = Array.isArray(bundlesAndFileTokens)
    ? bundlesAndFileTokens
    : [bundlesAndFileTokens];

  if (bundlesAndFileTokens.length === 0) {
    throw new AppError({ code: 'NoBundles' });
  }

  // Separate bundles from file tokens
  const [fileTokens, bundles] = partition(bundlesAndFileTokens, isString);

  // Get bundles from file tokens
  bundles.push(...getBundles(fileTokens));

  const results = await Promise.all(
    bundles.map(bundle =>
      exploreBundle(bundle, options).catch<ExploreErrorResult>(error =>
        onExploreError(bundle, error)
      )
    )
  );

  const exploreResult = getExploreResult(results, options);

  // Reject if none of results is successful
  if (exploreResult.bundles.length === 0) {
    return Promise.reject(exploreResult);
  }

  saveOutputToFile(exploreResult, options);

  return exploreResult;
}

/**
 * Expand list of file tokens into a list of bundles
 */
export function getBundles(fileTokens: string[]): Bundle[] {
  const filenames = flatMap(fileTokens, filePath =>
    glob.hasMagic(filePath) ? expandGlob(filePath) : filePath
  );

  const [mapFilenames, codeFilenames] = partition(filenames, filename => filename.endsWith('.map'));

  return codeFilenames.map<Bundle>(code => ({
    code,
    map: mapFilenames.find(filename => filename === `${code}.map`),
  }));
}

function expandGlob(pattern: string): string[] {
  // Make sure pattern match `.map` files as well
  if (pattern.endsWith('.js')) {
    pattern = `${pattern}?(.map)`;
  }

  return glob.sync(pattern);
}

export function getBundleName(bundle: Bundle): string {
  return Buffer.isBuffer(bundle.code) ? 'Buffer' : bundle.code;
}

/**
 * Handle error during bundle processing
 */
function onExploreError(bundle: Bundle, error: NodeJS.ErrnoException): ExploreErrorResult {
  return {
    bundleName: getBundleName(bundle),
    code: error.code || 'Unknown',
    message: error.message,
    error,
  };
}

function getExploreResult(
  results: (ExploreBundleResult | ExploreErrorResult)[],
  options: ExploreOptions
): ExploreResult {
  const [bundles, errors] = partition(
    results,
    (result): result is ExploreBundleResult => 'files' in result
  );

  errors.push(...getPostExploreErrors(bundles));

  return {
    bundles,
    errors,
    ...(bundles.length > 0 && { output: formatOutput(bundles, options) }),
  };
}

const SPECIAL_FILENAMES = [UNMAPPED_KEY, SOURCE_MAP_COMMENT_KEY];

function getPostExploreErrors(exploreBundleResults: ExploreBundleResult[]): ExploreErrorResult[] {
  const errors: ExploreErrorResult[] = [];

  const isSingleBundle = exploreBundleResults.length === 1;

  for (const result of exploreBundleResults) {
    const { bundleName, files, totalBytes } = result;

    // Check if source map contains only one file - this make result useless when exploring single bundle
    if (isSingleBundle) {
      const filenames = Object.keys(files).filter(
        filename => !SPECIAL_FILENAMES.includes(filename)
      );
      if (filenames.length === 1) {
        errors.push({
          bundleName,
          isWarning: true,
          code: 'OneSourceSourceMap',
          message: getErrorMessage({ code: 'OneSourceSourceMap', filename: filenames[0] }),
        });
      }
    }

    const unmappedBytes = files[UNMAPPED_KEY];
    if (unmappedBytes) {
      errors.push({
        bundleName,
        isWarning: true,
        code: 'UnmappedBytes',
        message: getErrorMessage({ code: 'UnmappedBytes', unmappedBytes, totalBytes }),
      });
    }
  }

  return errors;
}
