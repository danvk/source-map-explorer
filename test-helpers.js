const { existsSync } = require('fs');
const spawn = require('cross-spawn');
const concat = require('concat-stream');
const PATH = process.env.PATH;

/*
  `createProcess` and `execute` are from
  https://gist.github.com/zorrodg/c349cf54a3f6d0a9ba62e0f4066f31cb
  https://medium.com/@zorrodg/integration-tests-on-node-js-cli-part-1-why-and-how-fa5b1ba552fe
*/

/**
 * Creates a child process with script path
 * @param {string} processPath Path of the process to execute
 * @param {Array} args Arguments to the command
 */
function createProcess(processPath, args = []) {
  // Ensure that path exists
  if (!processPath || !existsSync(processPath)) {
    throw new Error('Invalid process path');
  }

  args = [processPath].concat(args);

  return spawn('node', args, {
    env: {
      NODE_ENV: 'test',
      preventAutoStart: false,
      PATH // This is needed in order to get all the binaries in your current terminal
    },
    stdio: [null, null, null, 'ipc'] // This enables interprocess communication (IPC)
  });
}

/**
 * Creates a command and executes inputs (user responses) to the stdin.
 * Returns a promise that resolves when all inputs are sent.
 * Rejects the promise if any error.
 * @param {string} processPath Path of the process to execute
 * @param {Array} args Arguments to the command
 */
function execute(processPath, args = []) {
  const childProcess = createProcess(processPath, args);
  childProcess.stdin.setEncoding('utf-8');
  const promise = new Promise((resolve, reject) => {
    childProcess.stderr.once('data', err => {
      reject(err.toString());
    });
    childProcess.on('error', reject);
    childProcess.stdout.pipe(
      concat(result => {
        resolve(result.toString());
      })
    );
  });
  return promise;
}

module.exports = {
  execute
};
