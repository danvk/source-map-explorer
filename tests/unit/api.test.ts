import { expect } from 'chai';
import fs from 'fs';
import rimraf from 'rimraf';
import snapshot from '@smpx/snap-shot-it';

import { explore, getBundles } from '../../src/lib/api';
import { setTestFolder, mockEOL } from '../test-helpers';

import type { BundlesAndFileTokens, ExploreOptions, Bundle } from '../../src/lib/types';

describe('api', () => {
  mockEOL();

  describe('explore', () => {
    setTestFolder();

    it('should generate data when provided with js file with inline map', async () => {
      const actual = await explore('data/inline-map.js');

      snapshot(actual);
    });

    it('should generate data when provided with file with referenced map', async () => {
      const actual = await explore('data/foo.min.js');

      snapshot(actual);
    });

    it('should generate data when provided with file with separated map file', async () => {
      const actual = await explore({
        code: 'data/no-map-comment.js',
        map: 'data/no-map-comment.js.map',
      });

      snapshot(actual);
    });

    it('should generate data respecting onlyMapped and replace options', async () => {
      const actual = await explore(
        { code: 'data/foo.min.js', map: 'data/foo.min.js.map' },
        {
          onlyMapped: true,
          replaceMap: { dist: 'hello' },
        }
      );

      snapshot(actual);
    });

    it('should generate data excluding source map bytes', async () => {
      const actual = await explore(
        ['data/inline-map.js', { code: 'data/foo.min.js', map: 'data/foo.min.js.map' }],
        {
          excludeSourceMapComment: true,
        }
      );

      snapshot(actual);
    });

    it('should generate data calculating gzip size', async () => {
      const actual = await explore(
        ['data/inline-map.js', { code: 'data/foo.min.js', map: 'data/foo.min.js.map' }],
        {
          gzip: true,
        }
      );

      snapshot(actual);
    });

    it('should sort bundles by name', async () => {
      const actual = await explore([
        'data/with-unmapped.js',
        'data/one-source.js',
        'data/null-source.js',
        'data/inline-map.js',
      ]);

      snapshot(actual);
    });

    it('should sort filenames', async () => {
      const actual = await explore('data/null-source.js', {
        sort: true,
      });

      snapshot(actual);
    });

    it('should accept buffer with inline map', async () => {
      const actual = await explore({ code: fs.readFileSync('data/inline-map.js') });

      snapshot(actual);
    });

    it('should accept buffers with js and map', async () => {
      const actual = await explore({
        code: fs.readFileSync('data/foo.min.js'),
        map: fs.readFileSync('data/foo.min.js.map'),
      });

      snapshot(actual);
    });

    it('should give name to "null" source a name', async () => {
      const actual = await explore('data/null-source.js');

      snapshot(actual);
    });

    describe('when output format specified', () => {
      it('should generate HTML from buffer', async () => {
        const result = await explore(
          { code: fs.readFileSync('data/inline-map.js') },
          { output: { format: 'html' } }
        );

        expect(result)
          .to.have.property('output')
          .that.contains('<title>Buffer - Source Map Explorer</title>')
          .and.contains('"bar.js')
          .and.contains('"foo.js');
      });

      it('should generate HTML from file', async () => {
        const result = await explore('data/foo.min.js', { output: { format: 'html' } });

        expect(result)
          .to.have.property('output')
          .that.contains('<title>data/foo.min.js - Source Map Explorer</title>')
          .and.contains('"bar.js')
          .and.contains('"foo.js');
      });

      it('should generate HTML from multiple files', async () => {
        const result = await explore(['data/foo.min.js', 'data/inline-map.js'], {
          output: { format: 'html' },
        });

        expect(result)
          .to.have.property('output')
          .that.contains('<title>[combined] - Source Map Explorer</title>')
          .and.contains('"bar.js')
          .and.contains('"foo.js');
      });

      it('should generate TSV', async () => {
        const result = await explore(['data/foo.min.js', 'data/inline-map.js'], {
          output: { format: 'tsv' },
        });

        snapshot(result.output);
      });

      it('should generate JSON', async () => {
        const result = await explore('data/foo.min.js', { output: { format: 'json' } });

        snapshot(result.output);
      });
    });

    describe(`when output filename specified`, () => {
      const tests: {
        name: string;
        filename: string;
        cleanPath?: string;
      }[] = [
        {
          name: 'should save html to file',
          filename: 'sme-result.html',
        },
        {
          name: 'should save html to file creating nested directories',
          filename: './tmp/nested/sme-result.html',
          cleanPath: './tmp',
        },
      ];

      tests.forEach(({ name, filename, cleanPath }) => {
        it(name, async () => {
          await explore('data/inline-map.js', {
            output: { format: 'html', filename },
          });

          expect(fs.existsSync(filename)).to.eq(true);

          const htmlContent = fs.readFileSync(filename).toString();

          expect(htmlContent).to.have.string(
            '<title>data/inline-map.js - Source Map Explorer</title>'
          );

          rimraf.sync(cleanPath || filename);
        });
      });
    });

    interface OnErrorTest {
      name: string;
      bundlesAndFileTokens: BundlesAndFileTokens;
      expectedErrorCode: string;
      options?: ExploreOptions;
    }

    describe('on error', () => {
      const bundleErrorTests: OnErrorTest[] = [
        {
          name: 'should throw when specified js file not found',
          bundlesAndFileTokens: 'data/something.js',
          expectedErrorCode: 'ENOENT',
        },
        {
          name: 'should throw when specified map file not found',
          bundlesAndFileTokens: { code: 'data/foo.min.js', map: 'data/foo.min.js.maap' },
          expectedErrorCode: 'ENOENT',
        },
        {
          name: 'should throw when cannot locate sourcemap',
          bundlesAndFileTokens: { code: 'data/no-map-comment.js' },
          expectedErrorCode: 'NoSourceMap',
        },
        {
          name: 'should throw if source map reference column beyond generated last column in line',
          bundlesAndFileTokens: 'data/invalid-map-column.js',
          expectedErrorCode: 'InvalidMappingColumn',
        },
        {
          name: 'should throw if source map reference more lines than available in source',
          bundlesAndFileTokens: 'data/invalid-map-line.js',
          expectedErrorCode: 'InvalidMappingLine',
        },
      ];

      bundleErrorTests.forEach(({ name, bundlesAndFileTokens, options, expectedErrorCode }) => {
        it(name, async () => {
          await expect(explore(bundlesAndFileTokens, options)).to.be.rejected.then((result) => {
            const error = result.errors[0];

            expect(error.code).to.equal(expectedErrorCode);
          });
        });
      });

      const appErrorTests: OnErrorTest[] = [
        {
          name: 'should throw when no bundles provided',
          bundlesAndFileTokens: [],
          expectedErrorCode: 'NoBundles',
        },
        {
          name: 'should throw when cannot save html to file',
          bundlesAndFileTokens: 'data/inline-map.js',
          // `/` is supposed to be invalid filename on both Linux and Windows
          options: { output: { format: 'html', filename: '/' } },
          expectedErrorCode: 'CannotSaveFile',
        },
      ];

      appErrorTests.forEach(({ name, bundlesAndFileTokens, options, expectedErrorCode }) => {
        it(name, async () => {
          await expect(explore(bundlesAndFileTokens, options)).to.be.rejected.then((error) => {
            expect(error.code).to.equal(expectedErrorCode);
          });
        });
      });

      it('should not throw if at least one result is successful', async () => {
        await expect(
          explore(['data/foo.min.js', 'data/no-map-comment.js'])
        ).to.not.be.rejected.then((result) => {
          expect(result.bundles.length).to.eq(1);
          expect(result.errors.length).to.eq(2);
        });
      });

      it('should not throw when analyzing source map referencing eol', async () => {
        await expect(explore('data/map-reference-eol.js')).to.not.be.rejected;
      });

      it('should not throw InvalidMappingColumn if noBorderChecks is true', async () => {
        await expect(explore('data/invalid-map-column.js', { noBorderChecks: true })).to.not.be
          .rejected;
      });

      it('should not throw InvalidMappingLine if noBorderChecks is true', async () => {
        await expect(explore('data/invalid-map-line.js', { noBorderChecks: true })).to.not.be
          .rejected;
      });

      it('should add "one source" source map warning when exploring single bundle', async () => {
        const result = await explore('data/one-source.js');

        const warning = result.errors[0];

        expect(warning.isWarning).to.equal(true);
        expect(warning.code).to.equal('OneSourceSourceMap');
      });

      it('should not add "one source" source map warning when exploring multiple bundles', async () => {
        const result = await explore(['data/one-source.js', 'data/inline-map.js'], {
          onlyMapped: true,
        });

        expect(result.errors).to.have.length(0);
      });

      it('should add warning about unmapped bytes', async () => {
        const result = await explore('data/with-unmapped.js');

        const warning = result.errors[1];

        expect(warning.isWarning).to.equal(true);
        expect(warning.code).to.equal('UnmappedBytes');
      });
    });
  });

  interface GetBundlesTest {
    name: string;
    fileTokens: string[];
    expected: Bundle[];
  }

  describe('getBundles', () => {
    setTestFolder();

    const tests: GetBundlesTest[] = [
      {
        name: 'should expand glob',
        fileTokens: ['data/foo.min.js*'],
        expected: [
          {
            code: 'data/foo.min.js',
            map: 'data/foo.min.js.map',
          },
        ],
      },
      {
        name: 'should group code and map file into bundle',
        fileTokens: ['data/foo.min.js', 'data/foo.min.js.map'],
        expected: [
          {
            code: 'data/foo.min.js',
            map: 'data/foo.min.js.map',
          },
        ],
      },
      {
        name: 'should expand glob into all bundles in directory',
        fileTokens: ['data/*.*'],
        expected: [
          {
            code: 'data/big.js',
            map: undefined,
          },
          {
            code: 'data/foo.min.js',
            map: 'data/foo.min.js.map',
          },
          {
            code: 'data/inline-map.js',
            map: undefined,
          },
          {
            code: 'data/invalid-map-column.js',
            map: undefined,
          },
          {
            code: 'data/invalid-map-line.js',
            map: undefined,
          },
          { code: 'data/map-reference-eol.js', map: undefined },
          {
            code: 'data/no-map-comment.js',
            map: 'data/no-map-comment.js.map',
          },
          { code: 'data/null-source.js', map: undefined },
          {
            code: 'data/one-source.js',
            map: undefined,
          },
          {
            code: 'data/with-unmapped.js',
            map: undefined,
          },
        ],
      },
      {
        name: 'should expand glob including .map files',
        fileTokens: ['data/foo.*.js'],
        expected: [
          {
            code: 'data/foo.min.js',
            map: 'data/foo.min.js.map',
          },
        ],
      },
      {
        name: 'should expand glob into code and map files',
        fileTokens: ['data/foo.*.js?(.map)'],
        expected: [
          {
            code: 'data/foo.min.js',
            map: 'data/foo.min.js.map',
          },
        ],
      },
      {
        name: 'should expand multiple globs',
        fileTokens: ['data/inline-*.js', 'data/foo.mi?.js'],
        expected: [
          {
            code: 'data/inline-map.js',
            map: undefined,
          },
          {
            code: 'data/foo.min.js',
            map: 'data/foo.min.js.map',
          },
        ],
      },
      {
        name: 'should support single file glob when inline map',
        fileTokens: ['data/inline*.js'],
        expected: [
          {
            code: 'data/inline-map.js',
            map: undefined,
          },
        ],
      },
    ];

    tests.forEach(({ name, fileTokens, expected }) => {
      it(name, () => {
        expect(getBundles(fileTokens)).to.deep.equal(expected);
      });
    });
  });
});
