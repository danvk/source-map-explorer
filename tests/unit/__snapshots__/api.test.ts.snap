exports['api explore should accept buffer with inline map 1'] = {
  "bundles": [
    {
      "bundleName": "Buffer",
      "totalBytes": 2999,
      "mappedBytes": 681,
      "unmappedBytes": 1,
      "eolBytes": 1,
      "sourceMapCommentBytes": 2316,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 480
        },
        "src/bar.js": {
          "size": 104
        },
        "src/foo.js": {
          "size": 97
        },
        "[sourceMappingURL]": {
          "size": 2316
        },
        "[unmapped]": {
          "size": 1
        },
        "[EOLs]": {
          "size": 1
        }
      }
    }
  ],
  "errors": [
    {
      "bundleName": "Buffer",
      "isWarning": true,
      "code": "UnmappedBytes",
      "message": "Unable to map 1/2999 bytes (0.03%)"
    }
  ]
}

exports['api explore should accept buffers with js and map 1'] = {
  "bundles": [
    {
      "bundleName": "Buffer",
      "totalBytes": 718,
      "mappedBytes": 681,
      "unmappedBytes": 1,
      "eolBytes": 1,
      "sourceMapCommentBytes": 35,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 480
        },
        "src/bar.js": {
          "size": 104
        },
        "src/foo.js": {
          "size": 97
        },
        "[sourceMappingURL]": {
          "size": 35
        },
        "[unmapped]": {
          "size": 1
        },
        "[EOLs]": {
          "size": 1
        }
      }
    }
  ],
  "errors": [
    {
      "bundleName": "Buffer",
      "isWarning": true,
      "code": "UnmappedBytes",
      "message": "Unable to map 1/718 bytes (0.14%)"
    }
  ]
}

exports['api explore should generate data calculating gzip size 1'] = {
  "bundles": [
    {
      "bundleName": "data/foo.min.js",
      "totalBytes": 392,
      "mappedBytes": 510,
      "eolBytes": 1,
      "sourceMapCommentBytes": 55,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 300
        },
        "src/bar.js": {
          "size": 108
        },
        "src/foo.js": {
          "size": 102
        },
        "[sourceMappingURL]": {
          "size": 55
        },
        "[EOLs]": {
          "size": 1
        }
      }
    },
    {
      "bundleName": "data/inline-map.js",
      "totalBytes": 1576,
      "mappedBytes": 510,
      "eolBytes": 1,
      "sourceMapCommentBytes": 1208,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 300
        },
        "src/bar.js": {
          "size": 108
        },
        "src/foo.js": {
          "size": 102
        },
        "[sourceMappingURL]": {
          "size": 1208
        },
        "[EOLs]": {
          "size": 1
        }
      }
    }
  ],
  "errors": []
}

exports['api explore should generate data excluding source map bytes 1'] = {
  "bundles": [
    {
      "bundleName": "data/foo.min.js",
      "totalBytes": 683,
      "mappedBytes": 681,
      "unmappedBytes": 1,
      "eolBytes": 1,
      "sourceMapCommentBytes": 35,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 480
        },
        "src/bar.js": {
          "size": 104
        },
        "src/foo.js": {
          "size": 97
        },
        "[unmapped]": {
          "size": 1
        },
        "[EOLs]": {
          "size": 1
        }
      }
    },
    {
      "bundleName": "data/inline-map.js",
      "totalBytes": 683,
      "mappedBytes": 681,
      "unmappedBytes": 1,
      "eolBytes": 1,
      "sourceMapCommentBytes": 2316,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 480
        },
        "src/bar.js": {
          "size": 104
        },
        "src/foo.js": {
          "size": 97
        },
        "[unmapped]": {
          "size": 1
        },
        "[EOLs]": {
          "size": 1
        }
      }
    }
  ],
  "errors": [
    {
      "bundleName": "data/foo.min.js",
      "isWarning": true,
      "code": "UnmappedBytes",
      "message": "Unable to map 1/683 bytes (0.15%)"
    },
    {
      "bundleName": "data/inline-map.js",
      "isWarning": true,
      "code": "UnmappedBytes",
      "message": "Unable to map 1/683 bytes (0.15%)"
    }
  ]
}

exports['api explore should generate data respecting onlyMapped and replace options 1'] = {
  "bundles": [
    {
      "bundleName": "data/foo.min.js",
      "totalBytes": 718,
      "mappedBytes": 681,
      "eolBytes": 1,
      "sourceMapCommentBytes": 35,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 480
        },
        "src/bar.js": {
          "size": 104
        },
        "src/foo.js": {
          "size": 97
        },
        "[sourceMappingURL]": {
          "size": 35
        },
        "[EOLs]": {
          "size": 1
        }
      }
    }
  ],
  "errors": []
}

exports['api explore should generate data when provided with file with referenced map 1'] = {
  "bundles": [
    {
      "bundleName": "data/foo.min.js",
      "totalBytes": 718,
      "mappedBytes": 681,
      "unmappedBytes": 1,
      "eolBytes": 1,
      "sourceMapCommentBytes": 35,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 480
        },
        "src/bar.js": {
          "size": 104
        },
        "src/foo.js": {
          "size": 97
        },
        "[sourceMappingURL]": {
          "size": 35
        },
        "[unmapped]": {
          "size": 1
        },
        "[EOLs]": {
          "size": 1
        }
      }
    }
  ],
  "errors": [
    {
      "bundleName": "data/foo.min.js",
      "isWarning": true,
      "code": "UnmappedBytes",
      "message": "Unable to map 1/718 bytes (0.14%)"
    }
  ]
}

exports['api explore should generate data when provided with file with separated map file 1'] = {
  "bundles": [
    {
      "bundleName": "data/no-map-comment.js",
      "totalBytes": 682,
      "mappedBytes": 681,
      "unmappedBytes": 1,
      "eolBytes": 0,
      "sourceMapCommentBytes": 0,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 480
        },
        "src/bar.js": {
          "size": 104
        },
        "src/foo.js": {
          "size": 97
        },
        "[sourceMappingURL]": {
          "size": 0
        },
        "[unmapped]": {
          "size": 1
        }
      }
    }
  ],
  "errors": [
    {
      "bundleName": "data/no-map-comment.js",
      "isWarning": true,
      "code": "UnmappedBytes",
      "message": "Unable to map 1/682 bytes (0.15%)"
    }
  ]
}

exports['api explore should generate data when provided with js file with inline map 1'] = {
  "bundles": [
    {
      "bundleName": "data/inline-map.js",
      "totalBytes": 2999,
      "mappedBytes": 681,
      "unmappedBytes": 1,
      "eolBytes": 1,
      "sourceMapCommentBytes": 2316,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 480
        },
        "src/bar.js": {
          "size": 104
        },
        "src/foo.js": {
          "size": 97
        },
        "[sourceMappingURL]": {
          "size": 2316
        },
        "[unmapped]": {
          "size": 1
        },
        "[EOLs]": {
          "size": 1
        }
      }
    }
  ],
  "errors": [
    {
      "bundleName": "data/inline-map.js",
      "isWarning": true,
      "code": "UnmappedBytes",
      "message": "Unable to map 1/2999 bytes (0.03%)"
    }
  ]
}

exports['api explore should give name to "null" source a name 1'] = {
  "bundles": [
    {
      "bundleName": "data/null-source.js",
      "totalBytes": 7856,
      "mappedBytes": 1006,
      "unmappedBytes": 13,
      "eolBytes": 1,
      "sourceMapCommentBytes": 6836,
      "files": {
        "webpack:///webpack/bootstrap": {
          "size": 897
        },
        "[no source]": {
          "size": 32
        },
        "webpack:///tests/generate-data/src/foo.js": {
          "size": 35
        },
        "webpack:///tests/generate-data/src/bar.js": {
          "size": 42
        },
        "[sourceMappingURL]": {
          "size": 6836
        },
        "[unmapped]": {
          "size": 13
        },
        "[EOLs]": {
          "size": 1
        }
      }
    }
  ],
  "errors": [
    {
      "bundleName": "data/null-source.js",
      "isWarning": true,
      "code": "UnmappedBytes",
      "message": "Unable to map 13/7856 bytes (0.17%)"
    }
  ]
}

exports['api explore when output format specified should generate JSON 1'] = `
{
  "results": [
    {
      "bundleName": "data/foo.min.js",
      "totalBytes": 718,
      "mappedBytes": 681,
      "unmappedBytes": 1,
      "eolBytes": 1,
      "sourceMapCommentBytes": 35,
      "files": {
        "../../node_modules/browser-pack/_prelude.js": {
          "size": 480
        },
        "src/bar.js": {
          "size": 104
        },
        "src/foo.js": {
          "size": 97
        },
        "[sourceMappingURL]": {
          "size": 35
        },
        "[unmapped]": {
          "size": 1
        },
        "[EOLs]": {
          "size": 1
        }
      }
    }
  ]
}
`

exports['api explore when output format specified should generate TSV 1'] = `
Source\tSize\r
../../node_modules/browser-pack/_prelude.js\t480\r
src/bar.js\t104\r
src/foo.js\t97\r
[sourceMappingURL]\t35\r
[unmapped]\t1\r
[EOLs]\t1\r
\r
[sourceMappingURL]\t2316\r
../../node_modules/browser-pack/_prelude.js\t480\r
src/bar.js\t104\r
src/foo.js\t97\r
[unmapped]\t1\r
[EOLs]\t1
`
